src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerFeature.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Http.Features.Internal;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    public class RequestServicesFeature : IServiceProvidersFeature, IDisposable
+    {
+        private IServiceProvider _requestServices;
+        private IServiceScope _scope;
+        private bool _requestServicesSet;
+
+        public RequestServicesFeature(IServiceProvider applicationServices)
+        {
+            if (applicationServices == null)
+            {
+                throw new ArgumentNullException(nameof(applicationServices));
+            }
+
+            ApplicationServices = applicationServices;
+        }
+
+        public IServiceProvider ApplicationServices { get; set; }
+
+        public IServiceProvider RequestServices
+        {
+            get
+            {
+                if (!_requestServicesSet)
+                {
+                    _scope = ApplicationServices.GetRequiredService<IServiceScopeFactory>().CreateScope();
```

Comment:
This behavior is totally undefined. To repro this you would have to set ApplicationServices to null then before ever touching RequestServices (I just know somebody is going to do something stupid and report a bug :smile:). This works today because nothing is lazy. We could store the original application services and use that as a fallback but that might be perplexing. Throwing might be the most reasonable thing. Null isn't bad either so lets flip a coin. If you did this as a customer an exception might be the best since you'd want to know why instead of getting a null ref. On the other hand if you get RequestServices and then set ApplicationServices to null, should that throw?